### PR TITLE
infoschema: add information_schema.tidb_cluster_info to retrieve cluster topology.

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <heng@lonng.org>

### What problem does this PR solve? 

In the current version, we don't have a simple way to get cluster information. We need to use different interfaces of different components or even different ways to get topology information.

### What is changed and how it works?

This PR introduces the `information_schema.tidb_cluster_info` and make the user can retrieve cluster topology easier via `select * from information_schema.tidb_cluster_info`.

### Check List 

Tests 

 - Unit test
 - Manual test
```
+------+------+--------+-----------------+-----------------+-----------------------------------------+------------------------------------------+
| ID   | TYPE | NAME   | ADDRESS         | STATUS_ADDRESS  | VERSION                                 | GIT_HASH                                 |
+------+------+--------+-----------------+-----------------+-----------------------------------------+------------------------------------------+
|    1 | tidb | tidb-0 | 127.0.0.1:4000  | 127.0.0.1:10080 | 5.7.25-TiDB-v4.0.0-alpha-666-g5b4111881 | 5b4111881ec4a693fdb417e415e4d87df4ed067e |
|    2 | pd   | pd-0   | 127.0.0.1:2379  | 127.0.0.1:2379  | 4.0.0-alpha                             | 8b2225d417f0991c3b11c09d2a2f7b6742e0ca78 |
|    3 | tikv | tikv-0 | 127.0.0.1:20160 | 127.0.0.1:20180 | 4.0.0-alpha                             | 48ac29e3c2302bd75cf2784f4c2f26ff5034db56 |
+------+------+--------+-----------------+-----------------+-----------------------------------------+------------------------------------------+
```

Release note

 - [feature] Add support for retrieving cluster topology via `select * from information_schema.tidb_cluster_info`

### CODE REVIEW

File: infoschema/tables.go [hunk 1]
```diff
@@ -1828,6 +1843,112 @@ func dataForServersInfo() ([][]types.Datum, error) {
 	return rows, nil
 }
 
+func dataForTiDBClusterInfo(ctx sessionctx.Context) ([][]types.Datum, error) {
+	// get TiDB servers info.
```
Comment (zimulala) [inline infoschema/tables.go]: s/get/Get

File: infoschema/tables.go [hunk 2]
```diff
@@ -1828,6 +1843,112 @@ func dataForServersInfo() ([][]types.Datum, error) {
...
+	store := ctx.GetStore()
+	etcd, ok := store.(tikv.EtcdBackend)
+	if !ok {
+		return nil, errors.Errorf("%T not an etcd backend", store)
+	}
+	for i, addr := range etcd.EtcdAddrs() {
+		addr = strings.TrimSpace(addr)
+
+		// get pd version
+		url := fmt.Sprintf("http://%s/pd/api/v1/config/cluster-version", addr)
```
Comment (zimulala) [inline infoschema/tables.go]: Could we use `ClusterVersion` here?

File: infoschema/tables.go [hunk 3]
```diff
@@ -1828,6 +1843,112 @@ func dataForServersInfo() ([][]types.Datum, error) {
...
+		}
+		pdVersion, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		terror.Log(resp.Body.Close())
+		version := strings.Trim(strings.Trim(string(pdVersion), "\n"), "\"")
+
+		// get pd git_hash
+		url = fmt.Sprintf("http://%s/pd/api/v1/status", addr)
```
Comment (zimulala) [inline infoschema/tables.go]: Could we use `Status`?

File: infoschema/tables.go [hunk 4]
```diff
@@ -1828,6 +1843,112 @@ func dataForServersInfo() ([][]types.Datum, error) {
...
+	// get PD servers info.
+	store := ctx.GetStore()
+	etcd, ok := store.(tikv.EtcdBackend)
+	if !ok {
+		return nil, errors.Errorf("%T not an etcd backend", store)
+	}
+	for i, addr := range etcd.EtcdAddrs() {
+		addr = strings.TrimSpace(addr)
+
+		// get pd version
```
Comment (zimulala) [inline infoschema/tables.go]: s/pd/PD

File: infoschema/tables.go (hunk 2 above)
Comment (djshow832) [inline infoschema/tables.go]: I suggest adding this url to `pdapi/const.go`.

File: infoschema/tables.go [hunk 5]
```diff
@@ -1828,6 +1843,112 @@ func dataForServersInfo() ([][]types.Datum, error) {
...
+		url := fmt.Sprintf("http://%s/pd/api/v1/config/cluster-version", addr)
+		resp, err := http.Get(url)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		pdVersion, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		terror.Log(resp.Body.Close())
```
Comment (djshow832) [inline infoschema/tables.go]: Why log?

Comment (lonng) [reply infoschema/tables.go]: `terror.Log` it a utility function equals `if err != nil {...}`.

**Comment by djshow832** [comment]
LGTM

File: infoschema/tables.go [hunk 6]
```diff
@@ -1828,6 +1843,112 @@ func dataForServersInfo() ([][]types.Datum, error) {
...
+		version := strings.Trim(strings.Trim(string(pdVersion), "\n"), "\"")
+
+		// Get PD git_hash
+		url = fmt.Sprintf("http://%s%s", addr, pdapi.Status)
+		resp, err = http.Get(url)
```
Comment (crazycs520) [inline infoschema/tables.go]: The request will always handle by PD leader, should also change PD?
image

Comment (lonng) [reply infoschema/tables.go]: Nice catch, seems PD should not redirect all request to leader.

Comment (lonng) [reply infoschema/tables.go]: https://github.com/pingcap/pd/pull/1883

**Pull request merged**